packages/blocks/src/components/virgo-input/virgo-input.ts

```diff
@@ -0,0 +1,211 @@
+            this.vEditor.yText.delete(vRange.index, vRange.length);
+          }
+          const length = this.vEditor.yText.length;
+          const restLength = this.maxLength - length;
+          if (restLength <= 0) {
+            return;
+          }
+          const text =
+            data.length > restLength ? data.slice(0, restLength) : data;
+          const originalText = this.vEditor.yText.toString();
+          const tmpText = `${originalText.substring(
+            0,
+            vRange.index
+          )}${text}${originalText.substring(vRange.index)}`;
+
+          if (this.type === 'number' && !isDecimal(tmpText)) {
+            return;
+          }
+
+          this.vEditor.insertText(vRange, text);
+          this.vEditor.setVRange({
+            index: vRange.index + text.length,
+            length: 0,
+          });
+          this.undoManager.stopCapturing();
+        }
+      },
+      virgoInput: ctx => {
+        const vRange = this.vEditor.getVRange();
+        if (!vRange) {
+          return ctx;
+        }
+        if (vRange.length > 0) {
+          this.vEditor.yText.delete(vRange.index, vRange.length);
+        }
+
+        const originalText = this.vEditor.yText.toString();
+        const tmpText = `${originalText.substring(0, vRange.index)}${
+          ctx.data
+        }${originalText.substring(vRange.index)}`;
+
+        let flag = true;
+
+        if (tmpText.length >= this.maxLength) {
+          ctx.skipDefault = true;
+          flag = false;
+        }
+
+        if (this.type === 'number' && !isDecimal(tmpText)) {
+          ctx.skipDefault = true;
+          flag = false;
+        }
+
+        if (flag) {
+          this.undoManager.stopCapturing();
+        }
+
+        return ctx;
+      },
+      virgoCompositionEnd: ctx => {
+        const vRange = this.vEditor.getVRange();
+        if (!vRange) {
+          return ctx;
+        }
+        if (vRange.length > 0) {
+          this.vEditor.yText.delete(vRange.index, vRange.length);
+        }
+
+        const originalText = this.vEditor.yText.toString();
+        const tmpText = `${originalText.substring(0, vRange.index)}${
+          ctx.data
+        }${originalText.substring(vRange.index)}`;
+
+        let flag = true;
+
+        if (tmpText.length >= this.maxLength) {
+          // We should not use `skipDefault` because we need to clear text node from IME.
+          ctx.data = '';
+          flag = false;
+        }
+
+        if (this.type === 'number' && !isDecimal(tmpText)) {
+          ctx.data = '';
+          flag = false;
+        }
+
+        if (flag) {
+          this.undoManager.stopCapturing();
+        }
+
+        return ctx;
+      },
+      keydown: e => {
+        if (
+          e instanceof KeyboardEvent &&
+          (e.ctrlKey || e.metaKey) &&
+          (e.key === 'z' || e.key === 'Z')
+        ) {
+          e.preventDefault();
+          if (e.shiftKey) {
+            this.undoManager.redo();
```

Comment:
```suggestion
            this.redo();
```
This is more concise.